app/models/git_repository.rb

```diff
@@ -155,6 +183,36 @@ def handling_worktree_errors
     raise MiqException::Error, err.message
   end
 
+  def git_lock_filename
+    @git_lock_filename ||= LOCKFILE_DIR.join(id.to_s)
+  end
+
+  def acquire_git_lock
+    return false if defined?(@git_lock) && !@git_lock.nil?
+
+    FileUtils.mkdir_p(LOCKFILE_DIR)
+
+    @git_lock = File.open(git_lock_filename, File::RDWR | File::CREAT, 0o644)
+    @git_lock.flock(File::LOCK_EX)                         # block waiting for lock
+    @git_lock.write("#{Process.pid} - #{Time.zone.now}\n") # for debugging
+    @git_lock.flush                                        # write current data
+    @git_lock.truncate(@git_lock.pos)                      # clean up remaining chars
```

Comment:
Why is this last line needed if you've flushed?

Reply:
Because if you don't, you could end up with something like this:

```ruby
File.open "tmp/fake.lock", File::RDWR | File::CREAT do |f|
  f.puts "really_really_long_pid - consistent_length_timestamp"
  f.flush
end

puts
puts "First Write"
puts "-----------"
puts
puts File.read "tmp/fake.lock"

File.open "tmp/fake.lock", File::RDWR | File::CREAT do |f|
  f.puts "short_pid - consistent_length_timestamp"
  f.flush
end

puts
puts "Second write"
puts "------------"
puts
puts File.read "tmp/fake.lock"


# $ ruby tmp/test_ruby_file_truncation.rb
#
# First Write
# -----------
#
# really_really_long_pid - consistent_length_timestamp
#
# Second write
# ------------
#
# short_pid - consistent_length_timestamp
# th_timestamp
```

Where the `"th_timestamp"` is the leftovers from the previous file write operation.

`File#flush` is there just to make sure we push to disk as soon as possible and not let it buffer in memory until we close the file (pretty sure), and `File#truncate` clears out any data after the cursor position (in this case).

Reply:
Wow that is funky.  I would have expected the flush to ensure that doesn't happen.  Another option is to use `f.sync = true` maybe?

Reply:
Ohhhhh I see...it took me a while to understand but this opens an existing file for random access, as opposed to the usual open a file for write and auto-truncate.